enhancements/dns/coredns-egressfirewall-integration.md

```diff
@@ -0,0 +1,505 @@
+type EgressFirewallDNSName struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// metadata is the standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// spec is the specification of the desired behavior of the EgressFirewallDNSName.
+	Spec EgressFirewallDNSNameSpec `json:"spec,omitempty"`
+	// status is the most recently observed status of the EgressFirewallDNSName.
+	Status EgressFirewallDNSNameStatus `json:"status,omitempty"`
+}
+
+// EgressFirewallDNSNameSpec is a desired state description of EgressFirewallDNSName.
+type EgressFirewallDNSNameSpec struct {
+	// name is the DNS name used in a EgressFirewall rule.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Pattern=^(\*\.)?([A-Za-z0-9-]+\.)*[A-Za-z0-9-]+\.$
+	Name string `json:"name"`
+}
+
+// EgressFirewallDNSNameStatus defines the observed status of EgressFirewallDNSName.
+type EgressFirewallDNSNameStatus struct {
+	// The list of matching DNS names and their corresponding IP addresses along with TTL and last
+	// DNS lookup time.
+	ResolvedNames []EgressFirewallDNSNameStatusItem `json:"resolvedNames,omitempty"`
+}
+
+// EgressFirewallDNSNameStatusItem describes the details of a resolved DNS name.
+type EgressFirewallDNSNameStatusItem struct {
+	// The resolved DNS name corresponding to the Name field of EgressFirewallDNSNameSpec.
+	// +kubebuilder:validation:Pattern=^(\*\.)?([A-Za-z0-9-]+\.)*[A-Za-z0-9-]+\.$
+	DNSName string `json:"dnsName"`
+	// The IP addresses associated with the DNS name used in a EgressFirewall rule.
+	IPs []string `json:"ips"`
+	// Minimum time-to-live value among all the IP addresses.
+	TTL int64 `json:"ttl"`
+	// Timestamp when the last DNS lookup was successfully completed.
+	LastLookupTime metav1.Time `json:"lastLookupTime"`
+}
+````
+
+### Workflow Description
+
+The workflows for Create, Delete and Update events for EgressFirewall related to DNS names are explained in this section. The workflow
+for those events is shown in the following diagram with an example `EgressFirewallDNSName` CR:
+
+![Workflow](./coredns-egressfirewall-integration.png)
+
+#### Create/Update of DNS name
+
+* An OpenShift cluster administrator creates/updates an EgressFirewall resource for a namespace and adds rule(s) containing
+DNS name(s).
+* The OVN-K cluster manager will create corresponding `EgressFirewallDNSName` CRs for each of the DNS names in the EgressFirewall rules, if not
+already created. Each CR will be created in the `openshift-ovn-kubernetes` namespace. The name of the CR will be assigned using a hash
+function (similar to the ComputeHash [here](https://github.com/openshift/kubernetes/blob/master/pkg/controller/controller_utils.go#L1157-L1172))
+prefixed by `dns-`. The `.spec.name` field of the CR will be set to the DNS name (along with a trailing `.`).
```

Comment:
DNS servers store DNS names with a trailing period and whenever a DNS request is sent to CoreDNS, the trailing period is added, if it's not already added. So, DNS name matching during lookup will be just simple string matching. 

Additionally, when the status will be updated by CoreDNS, it will add the DNS names with the trailing period. So, matching the DNS names in the status also will be simple string matching.

Reply:
This context would be good to include here, I didn't understand why this was being done either.
From a UX perspective though, I would expect the values to match between the user input rule and the name here, why not do the trailing dot addition inside the plugin itself?

Reply:
The CR is not for the users actually. The whole lifecycle is through OCP components without any user intervention.

Also, as only a single EgressFirewallDNSName CR will be created for an unique DNS name, even if the DNS name is used in multiple EgressFirewall objects, there is no guarantee how the user adds the DNS name (with or without trailing period) in each of the EgressFirewall DNS rules.

Thus, to make it consistent, and also for ease of matching, the `spec.name` will be set with a trailing period.